library/linux/src/internal/json_message_codec.h

```diff
@@ -0,0 +1,59 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef LIBRARY_LINUX_INCLUDE_FLUTTER_DESKTOP_EMBEDDING_JSON_MESSAGE_CODEC_H_
+#define LIBRARY_LINUX_INCLUDE_FLUTTER_DESKTOP_EMBEDDING_JSON_MESSAGE_CODEC_H_
+
+#include <memory>
+#include <vector>
+
+#include <json/json.h>
+
+namespace flutter_desktop_embedding {
+
+// A message encoding/decoding mechanism for communications to/from the
+// Flutter engine via JSON channels.
+//
+// TODO: Make this public, once generalizing a MessageCodec parent interface is
+// addressed; this is complicated by the return type of EncodeMessage.
+// Part of issue #102.
+class JsonMessageCodec {
+ public:
+  // Returns the shared instance of the codec.
+  static const JsonMessageCodec &GetInstance();
+
+  ~JsonMessageCodec() = default;
+
+  // Prevent copying.
+  JsonMessageCodec(JsonMessageCodec const &) = delete;
+  JsonMessageCodec &operator=(JsonMessageCodec const &) = delete;
+
+  // Returns a binary encoding of the given message, or nullptr if the
+  // message cannot be serialized by this codec.
+  std::unique_ptr<std::vector<uint8_t>> EncodeMessage(
+      const Json::Value &message) const;
+
+  // Returns the MethodCall encoded in |message|, or nullptr if it cannot be
```

Comment:
Sorry, this was a copypasta. It should have been s/MethodCall/message/. I'll fix and re-upload tomorrow.

For more context, the Flutter APIs that I'm moving toward following have:

- MessageChannel: A named channel for communicating between Flutter and platform code. A MessageChannel has a specific codec associated with it at creation (JSON, Standard, etc.) but beyond that doesn't have any expectations about the contents, so clients can define whatever structure they want.
- MethodChannel: A layer on top of MessageChannel which adds a few expectations about structure: the incoming message should have a method name and optional arguments, and there should be a response that communicates success/error/not-implemented using a defined structure.

The Codec family of classes has the same split, with the MethodCodec family knowing about encoding/decoding that added layer of structure.

Our existing Plugin API doesn't have this distinction, because everything we built it to interface with so far has used MethodChannel on the Flutter side. But we do want to support it in general as part of the API alignment I'm working on, and in particular it's relevant now because the raw key event handling code in Flutter uses MessageChannel rather than MethodChannel.

So this is a small step toward adding that separation in the desktop APIs as well.

> If it is the case that this is supposed to be more general, I might also ask that beyond updating this comment, to update the class header as well. It seems this is just a generally JSON message decoder rather than one that verifies any codec.

In the Flutter API terminology, a (Message)Codec is just an encoder/decoder. It's not until you get to the MethodCodec layer that there's an assumed structure beyond how different data types are represented. I could make the class comment slightly more generic by removing the reference to channels, but the reason the Codec classes exist as an API is for use by channels; if someone wanted to do JSON encoding/decoding for reasons other than channel messages they should just use an existing JSON API (e.g., jsoncpp) directly.